src/SlimBootstrap/ResponseOutputWriter/Csv.php

```diff
@@ -0,0 +1,394 @@
+     * @param String             $shortName
+     */
+    public function __construct(
+        Slim\Http\Request $request,
+        Slim\Http\Response $response,
+        Slim\Http\Headers $headers,
+        $shortName
+    ) {
+        $this->_request  = $request;
+        $this->_response = $response;
+        $this->_headers  = $headers;
+    }
+
+    /**
+     * This function outputs the given $data as valid CSV to the client
+     * and sets the HTTP Response Code to the given $statusCode.
+     *
+     * @param SlimBootstrap\DataObject[] $data The data to output to
+     *                                                   the client
+     * @param int $statusCode The status code to set
+     *                                                   in the response
+     * @throws CSVEncodingException
+     */
+    public function write($data, $statusCode = 200)
+    {
+        $result = array();
+
+        if (true === \is_array($data)) {
+            $data = $this->_normalizeAll($data);
+            $identifiers = null;
+            foreach ($data as $entry) {
+                $this->_buildStructure(
+                    $entry,
+                    $entry->getIdentifiers(),
+                    0,
+                    $result
+                );
+            }
+        } else if ($data instanceof DataObject) {
+            $this->_buildStructure(
+                $this->_normalizeOne($data),
+                $data->getIdentifiers(),
+                0,
+                $result
+            );
+        } else {
+            throw new CSVEncodingException(
+                "Expected DataObject, " . \gettype($data) . " given."
+            );
+        }
+
+        $body = $this->_csvEncode($result);
+
+//        if (false === $body) {
+//            $this->_response->setStatus(500);
+//            $this->_response->setBody("Error encoding requested data.");
+//            return;
+//        }
+
+        $this->_headers->set(
+            'Content-Type',
+            'text/csv; charset=UTF-8'
+        );
+
+        $this->_response->setStatus($statusCode);
+        $this->_response->setBody($body);
+    }
+
+    /**
+     * This function adds keys to the DataObject that are necessary
+     * to form uniform data. Nonexistent keys will be filled with null.
+     *
+     * @param SlimBootstrap\DataObject  $data   DataObject to modify
+     * @param array|null                $keys   Array of keys that need
+     *                                          to exist.
+     * @return SlimBootstrap\DataObject         DataObject w/ filled Keys
+     */
+    protected function _normalizeOne(DataObject $data, $keys = null){
+        if(null === $keys){
+            $keys = \array_keys($data->getData());
+        }
+
+        $keys = \array_fill_keys($keys, null);
+
+        if (\count(
+                \array_intersect_key(
+                    $keys,
+                    $data->getData()
```

Comment:
please move count call out of the if condition, so the condition is better readable